{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c3adb4a-26f5-4e47-a8f8-2872de880004",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scipy.sparse as sparse\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import demes, msprime, demesdraw\n",
    "import tspop, tskit, tstrait\n",
    "import cython\n",
    "import statsmodels.api as sm\n",
    "%load_ext cython\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arial\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccb31e8e-8f58-45ff-a2ea-45c9a82ce342",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%cython --cplus\n",
    "\n",
    "import numpy as np\n",
    "cimport numpy as np\n",
    "cimport cython\n",
    "\n",
    "@cython.boundscheck(False)\n",
    "def get_lanc_dense(\n",
    "    const long[:] seg_left, # left of segment\n",
    "    const long[:] seg_right, # right of segment\n",
    "    const long[:] seg_lanc, # lanc of segment\n",
    "    const long[:] node_ptr, # similar to sparse mat indptr\n",
    "    const long[:] site_pos, # position of sites\n",
    "    int n_site, # number of sample nodes\n",
    "    int n_node # number of sites\n",
    "):\n",
    "    \n",
    "    cdef long[::1,:] lanc_dense = np.empty((n_site, n_node), dtype=long, order='F')\n",
    "\n",
    "    cdef Py_ssize_t i_node, i_site, \n",
    "    cdef Py_ssize_t node_begin, node_end, pos_site\n",
    "    cdef Py_ssize_t seg_begin, seg_end\n",
    "    with nogil:\n",
    "        for i_node in range(n_node):\n",
    "            node_begin, node_end = node_ptr[i_node], node_ptr[i_node+1]\n",
    "            for i_site in range(n_site):\n",
    "                pos_site = site_pos[i_site]\n",
    "                for i_seg in range(node_begin, node_end):\n",
    "                    seg_begin, seg_end = seg_left[i_seg], seg_right[i_seg]\n",
    "                    if (pos_site >= seg_begin) and (pos_site < seg_end):\n",
    "                        lanc_dense[i_site, i_node] = seg_lanc[i_seg]\n",
    "\n",
    "    return np.asarray(lanc_dense)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cdeb1c8-2b1b-46f1-a0f7-9833c3aeae5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://github.com/tskit-dev/tskit/issues/504\n",
    "def allele_frequencies(ts, sample_sets=None):\n",
    "    if sample_sets is None:\n",
    "       sample_sets = [ts.samples()] \n",
    "    n = np.array([len(x) for x in sample_sets])\n",
    "    def f(x):\n",
    "       return x / n\n",
    "    return ts.sample_count_stat(sample_sets, f, len(sample_sets), windows='sites', polarised=True, mode='site', strict=False, span_normalise=False)\n",
    "\n",
    "# convert site x node to site x individual matrix\n",
    "def hap_to_dip(mat): # assumes diploid\n",
    "    n_site, n_node = mat.shape\n",
    "    n_individual = int(n_node/2)\n",
    "\n",
    "    data = np.ones(n_node)\n",
    "    indices = np.repeat(np.arange(n_individual),2)\n",
    "    indptr = np.arange(n_node+1)\n",
    "    \n",
    "    sum_mat = sparse.csr_matrix((data,indices,indptr), shape=(n_node, n_individual))\n",
    "    return(mat @ sum_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1973a0e3-f1d0-4fb0-be30-7c7f67ed1d79",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = demes.load('amr_demes.yml')\n",
    "ax = demesdraw.tubes(model, log_time=True)\n",
    "plt.savefig('history.svg')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "212997f1-bf70-41dd-aaf3-e63160fa9352",
   "metadata": {},
   "outputs": [],
   "source": [
    "demography = msprime.Demography.from_demes(model)\n",
    "# 12 is start_time=300 divided by generation time=25\n",
    "demography.add_census(time=12.01) # census required to track ancestors\n",
    "demography.sort_events() # event should be sorted after adding census\n",
    "samples = {'AFR':5000, 'EUR':5000, 'AMR': 10000}\n",
    "# rates from https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html#sec_catalog_HomSap\n",
    "# chromosome 22\n",
    "seq_length = 50818468\n",
    "ts = msprime.sim_ancestry(\n",
    "    [msprime.SampleSet(n, ploidy=2, population=p) for p, n in samples.items()],\n",
    "    demography=demography,\n",
    "    model=[\n",
    "        msprime.DiscreteTimeWrightFisher(duration=5),\n",
    "        msprime.StandardCoalescent(),\n",
    "    ],\n",
    "    recombination_rate=2.10572e-08,\n",
    "    sequence_length=seq_length,\n",
    "    record_migrations=True,\n",
    "    random_seed=1\n",
    ")\n",
    "mts = msprime.sim_mutations(ts, rate=1.29e-08, random_seed=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32c7a6de-b3cf-4e40-a172-2c35becd74fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "af = allele_frequencies(mts, [mts.samples(0), mts.samples(1)])\n",
    "af_afr, af_eur = af[:,0], af[:,1]\n",
    "af_threshold = 0.01\n",
    "rare_variants = (af_afr < af_threshold) | (af_eur < af_threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35307d0a-d977-469a-9f1b-554cb5b95daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "mts_common = mts.delete_sites(np.arange(mts.num_sites)[rare_variants])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3b3eeb9-ba3d-4497-83d6-108a254c12ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_afr = mts_common.genotype_matrix(samples=mts_common.samples(0))\n",
    "gt_eur = mts_common.genotype_matrix(samples=mts_common.samples(1))\n",
    "gt_amr = mts_common.genotype_matrix(samples=mts_common.samples(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2599fac6-5c4b-4863-afab-cf6e13fd6450",
   "metadata": {},
   "outputs": [],
   "source": [
    "pa = tspop.get_pop_ancestry(mts_common, census_time=12.01) # admixture happend in t=12, census time t=10.01\n",
    "pa_amr = pa.subset_tables(subset_samples=mts_common.samples(2))[1] # subset AMR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98ac820c-94c3-4dcd-98cb-f35f49b9a00f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "seg_left = pa_amr.left.values.astype(int)\n",
    "seg_right = pa_amr.right.values.astype(int)\n",
    "seg_lanc = pa_amr.population.values.astype(int)\n",
    "node_ptr = np.zeros(gt_amr.shape[1]+1, dtype=int) # similar to sparse matrix pointer\n",
    "node_ptr[1:] = pa_amr['sample'].value_counts(sort=False).cumsum().values\n",
    "site_pos = mts_common.tables.sites.position.astype(int)\n",
    "n_site, n_node = gt_amr.shape\n",
    "\n",
    "lanc_dense = get_lanc_dense(seg_left, seg_right, seg_lanc, node_ptr, site_pos, n_site, n_node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a02a5824-e275-47e4-86c1-1792bc12167c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_amr_afr = gt_amr * (1-lanc_dense) \n",
    "gt_amr_eur = gt_amr * lanc_dense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae0a32f8-d787-4832-9a37-22bc040aa2bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_afr_dip = hap_to_dip(gt_afr)\n",
    "gt_eur_dip = hap_to_dip(gt_eur)\n",
    "gt_amr_dip = hap_to_dip(gt_amr)\n",
    "\n",
    "gt_amr_afr_dip = hap_to_dip(gt_amr_afr) \n",
    "gt_amr_eur_dip = hap_to_dip(gt_amr_eur)\n",
    "\n",
    "lanc_dip = hap_to_dip(lanc_dense)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae16301a-3f2b-4edd-ba74-b3195edc0dee",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_sites = mts_common.num_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f411f6e-d3f8-45d6-ae58-245856beae1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "polygenicity_props = [1e-3, 5e-3, 1e-2, 5e-2]\n",
    "\n",
    "afr_list = []\n",
    "eur_list = []\n",
    "amr_afr_list = [] \n",
    "amr_eur_list = []\n",
    "sim_result_list = []\n",
    "for polygenicity_prop in polygenicity_props:\n",
    "    \n",
    "    # simulate trait\n",
    "    trait_model = tstrait.trait_model(distribution=\"normal\", mean=0, var=1)\n",
    "    sim_result = tstrait.sim_phenotype(\n",
    "      ts=mts_common, num_causal=int(num_sites * polygenicity_prop), model=trait_model, h2=0.5, random_seed=1\n",
    "    )\n",
    "    sim_result_list.append(sim_result)\n",
    "    \n",
    "    pheno_afr = sim_result.phenotype.iloc[:5000,:]\n",
    "    pheno_eur = sim_result.phenotype.iloc[5000:10000,:]\n",
    "    pheno_amr = sim_result.phenotype.iloc[10000:,:]\n",
    "\n",
    "    # get true estimate \n",
    "    beta_afr = np.zeros(num_sites)\n",
    "    beta_eur = np.zeros(num_sites)\n",
    "    beta_amr_afr = np.zeros(num_sites)\n",
    "    beta_amr_eur = np.zeros(num_sites)\n",
    "    \n",
    "    for i in range(num_sites):\n",
    "        # afr\n",
    "        exog_afr = gt_afr_dip[i,:]\n",
    "        endog_afr = pheno_afr.genetic_value.values\n",
    "        fit_afr = sm.OLS(endog_afr, sm.add_constant(exog_afr)).fit().params\n",
    "        if len(fit_afr) > 1:\n",
    "            beta_afr[i] = fit_afr[1]\n",
    "        \n",
    "        # eur\n",
    "        exog_eur = gt_eur_dip[i,:]\n",
    "        endog_eur = pheno_eur.genetic_value.values\n",
    "        fit_eur = sm.OLS(endog_eur, sm.add_constant(exog_eur)).fit().params\n",
    "        if len(fit_eur) > 1:\n",
    "            beta_eur[i] = fit_eur[1]\n",
    "            \n",
    "        # amr\n",
    "        exog_afr = gt_amr_afr_dip[i,:]\n",
    "        exog_eur = gt_amr_eur_dip[i,:]\n",
    "        lanc = lanc_dip[i,:]\n",
    "        exog = np.vstack((exog_afr, exog_eur, lanc)).T\n",
    "        endog = pheno_amr.genetic_value.values\n",
    "        fit = sm.OLS(endog, sm.add_constant(exog)).fit().params\n",
    "        beta_amr_afr[i] = fit[1]\n",
    "        beta_amr_eur[i] = fit[2]\n",
    "\n",
    "    afr_list.append(beta_afr)\n",
    "    eur_list.append(beta_eur)\n",
    "    amr_afr_list.append(beta_amr_afr)\n",
    "    amr_eur_list.append(beta_amr_eur)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe406a45-4297-484d-acc8-981c67aa33d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "fig, ax = plt.subplots(2, 4, figsize=(17,8))\n",
    "\n",
    "for col in range(4):\n",
    "\n",
    "    site_causal = np.random.binomial(n=1, p=0.2, size=num_sites).astype(bool)\n",
    "    \n",
    "    x, y = afr_list[col][site_causal], amr_afr_list[col][site_causal]\n",
    "    xy = np.vstack([x,y])\n",
    "    z = gaussian_kde(xy)(xy)\n",
    "    ax[0, col].scatter(x, y, c=z, s=5, rasterized=True)\n",
    "\n",
    "    x, y = eur_list[col][site_causal], amr_eur_list[col][site_causal]\n",
    "    xy = np.vstack([x,y])\n",
    "    z = gaussian_kde(xy)(xy)\n",
    "    ax[1, col].scatter(x, y, c=z, s=5, rasterized=True)\n",
    "\n",
    "    lim= 2e-3 / np.sqrt(polygenicity_props[col]) * 1.1\n",
    "    ax[0,col].set_xlim([-lim,lim])\n",
    "    ax[0,col].set_ylim([-lim,lim])\n",
    "    ax[1,col].set_xlim([-lim,lim])\n",
    "    ax[1,col].set_ylim([-lim,lim])\n",
    "    ax[0,col].plot([-1,1],[-1,1],ls='--', color='black')\n",
    "    ax[1,col].plot([-1,1],[-1,1],ls='--', color='black')\n",
    "\n",
    "\n",
    "    afr_cor = np.corrcoef(afr_list[col], amr_afr_list[col])[0,1]\n",
    "    eur_cor = np.corrcoef(eur_list[col], amr_eur_list[col])[0,1]\n",
    "    ax[0,col].text(0.1, 0.9, r'$r_{g,all} = %.2f$' % afr_cor, fontsize=15, transform=ax[0,col].transAxes)\n",
    "    ax[1,col].text(0.1, 0.9, r'$r_{g,all} = %.2f$' % eur_cor, fontsize=15, transform=ax[1,col].transAxes)\n",
    "\n",
    "    z = np.log10(polygenicity_props[col])\n",
    "    pow, val = math.floor(z), int(10**(z-math.floor(z)) + 0.5)\n",
    "    ax[0,col].set_title(r'Causal proportion: $%d \\times 10^{%d}$' % (val, pow), fontsize=15)\n",
    "    \n",
    "    ax[0,col].locator_params(axis='both', nbins=5)\n",
    "    ax[1,col].locator_params(axis='both', nbins=5)\n",
    "\n",
    "ax[0,0].set_ylabel('African', fontsize=15)\n",
    "ax[1,0].set_ylabel('European', fontsize=15)\n",
    "\n",
    "fig.text(0.5, -0.02, 'Single-continental marginal effect size', fontsize=15, ha='center', va='center')\n",
    "fig.text(-0.02, 0.5, 'Admixed TRACTOR effect size', fontsize=15, ha='center', va='center', rotation=90)\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('full.svg', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd0641a4-1114-419b-9c9f-5e6f407a9d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"mathtext.fontset\"] = 'cm'\n",
    "\n",
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "fig, ax = plt.subplots(2, 4, figsize=(17,8))\n",
    "\n",
    "for col in range(4):\n",
    "\n",
    "    site_causal = sim_result_list[col].effect_size.site_id\n",
    "    \n",
    "    x, y = afr_list[col][site_causal], amr_afr_list[col][site_causal]\n",
    "    xy = np.vstack([x,y])\n",
    "    z = gaussian_kde(xy)(xy)\n",
    "    ax[0, col].scatter(x, y, c=z, s=10, rasterized=True)\n",
    "\n",
    "    x, y = eur_list[col][site_causal], amr_eur_list[col][site_causal]\n",
    "    xy = np.vstack([x,y])\n",
    "    z = gaussian_kde(xy)(xy)\n",
    "    ax[1, col].scatter(x, y, c=z, s=10, rasterized=True)\n",
    "\n",
    "    lim= 2e-3 / np.sqrt(polygenicity_props[col]) * 1.1\n",
    "    ax[0,col].set_xlim([-lim,lim])\n",
    "    ax[0,col].set_ylim([-lim,lim])\n",
    "    ax[1,col].set_xlim([-lim,lim])\n",
    "    ax[1,col].set_ylim([-lim,lim])\n",
    "    ax[0,col].plot([-1,1],[-1,1],ls='--', color='black')\n",
    "    ax[1,col].plot([-1,1],[-1,1],ls='--', color='black')\n",
    "\n",
    "\n",
    "    afr_cor = np.corrcoef(afr_list[col][site_causal], amr_afr_list[col][site_causal])[0,1]\n",
    "    eur_cor = np.corrcoef(eur_list[col][site_causal], amr_eur_list[col][site_causal])[0,1]\n",
    "    ax[0,col].text(0.1, 0.9, r'$r_{g,causal} = %.2f$' % afr_cor, fontsize=15, transform=ax[0,col].transAxes)\n",
    "    ax[1,col].text(0.1, 0.9, r'$r_{g,causal} = %.2f$' % eur_cor, fontsize=15, transform=ax[1,col].transAxes)\n",
    "\n",
    "    z = np.log10(polygenicity_props[col])\n",
    "    pow, val = math.floor(z), int(10**(z-math.floor(z)) + 0.5)\n",
    "    ax[0,col].set_title(r'Causal proportion: $%d \\times 10^{%d}$' % (val, pow), fontsize=15)\n",
    "\n",
    "    ax[0,col].locator_params(axis='both', nbins=5)\n",
    "    ax[1,col].locator_params(axis='both', nbins=5)\n",
    "\n",
    "ax[0,0].set_ylabel('African', fontsize=15)\n",
    "ax[1,0].set_ylabel('European', fontsize=15)\n",
    "\n",
    "fig.text(0.5, -0.02, 'Single-continental marginal effect size', fontsize=15, ha='center', va='center')\n",
    "fig.text(-0.02, 0.5, 'Admixed TRACTOR effect size', fontsize=15, ha='center', va='center', rotation=90)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('causal.svg', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00b7d18c-9387-477a-b46b-1a11f71ae440",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef_afr = np.zeros(num_sites)\n",
    "coef_eur = np.zeros(num_sites)\n",
    "for i in range(num_sites):\n",
    "    lanc = lanc_dip[i,:]\n",
    "    endog_afr = gt_amr_afr_dip[i,:]\n",
    "    endog_eur = gt_amr_eur_dip[i,:]\n",
    "\n",
    "    coef_afr[i] = np.dot(2-lanc, endog_afr) / np.dot(2-lanc,2-lanc)\n",
    "    coef_eur[i] = np.dot(lanc, endog_eur) / np.dot(lanc,lanc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1c74912-be84-4f38-b505-ef3eba2f8d38",
   "metadata": {},
   "outputs": [],
   "source": [
    "af_common = allele_frequencies(mts_common, [mts_common.samples(0), mts_common.samples(1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40f09661-a887-4777-9689-cf7158eeb4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "afr_resid = gt_amr_afr_dip - (2-lanc_dip) * coef_afr[:,None]\n",
    "eur_resid = gt_amr_eur_dip - lanc_dip * coef_eur[:,None]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bff88538-28c9-41be-b4e5-acff3f3c7f3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "n_loci = 5000\n",
    "site_sub = np.linspace(0, 69700, n_loci).astype(int)\n",
    "afr_resid_sub = afr_resid[site_sub,:]\n",
    "eur_resid_sub = eur_resid[site_sub,:]\n",
    "afr_sub = gt_afr_dip[site_sub,:]\n",
    "eur_sub = gt_eur_dip[site_sub,:]\n",
    "gt_amr_afr_sub = gt_amr_afr_dip[site_sub,:]\n",
    "gt_amr_eur_sub = gt_amr_eur_dip[site_sub,:]\n",
    "site_distance = squareform(pdist(mts_common.tables.sites[site_sub].position[:,None], 'cityblock'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56b67d98-0244-41c5-b176-3c0b0f888d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "afr_ld = np.nan_to_num(1/10000 * (afr_resid_sub @ gt_amr_afr_sub.T) / np.outer(afr_resid_sub.std(axis=1), gt_amr_afr_sub.std(axis=1)))\n",
    "eur_ld = np.nan_to_num(1/10000 * (eur_resid_sub @ gt_amr_eur_sub.T) / np.outer(eur_resid_sub.std(axis=1), gt_amr_eur_sub.std(axis=1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5e1e5f9-20fa-4a54-89a1-4ab2dd18887c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 2, figsize=(8.5,4), sharex=True, sharey=True)\n",
    "\n",
    "pt_plot = np.random.binomial(n=1,p=1e-1,size=n_loci*n_loci).astype(bool)\n",
    "im1 = ax[0].scatter(np.nan_to_num(np.corrcoef(afr_sub).ravel()[pt_plot]), afr_ld.ravel()[pt_plot], s=0.5, c=np.log10(site_distance.ravel()/seq_length)[pt_plot], rasterized=True, vmin=-4.5, vmax=-0.5)\n",
    "im2 = ax[1].scatter(np.nan_to_num(np.corrcoef(eur_sub).ravel()[pt_plot]), eur_ld.ravel()[pt_plot], s=0.5, c=np.log10(site_distance.ravel()/seq_length)[pt_plot], rasterized=True, vmin=-4.5, vmax=-0.5)\n",
    "\n",
    "ax[0].plot([-1,1],[-1,1],ls='--',c='black')\n",
    "ax[1].plot([-1,1],[-1,1],ls='--',c='black')\n",
    "\n",
    "lim = 0.6 * 1.05\n",
    "ax[0].set_xlim([-lim,lim])\n",
    "ax[0].set_ylim([-lim,lim])\n",
    "\n",
    "fig.text(0.5, -0.02, 'Single-continental LD correlation', fontsize=15, ha='center', va='center')\n",
    "fig.text(-0.02, 0.5, 'Local ancestry-adjusted LD correlation', fontsize=15, ha='center', va='center', rotation=90)\n",
    "\n",
    "ax[0].set_title('African', fontsize=15)\n",
    "ax[1].set_title('European', fontsize=15)\n",
    "\n",
    "afr_cor = np.corrcoef(np.nan_to_num(np.corrcoef(afr_sub).ravel()), afr_ld.ravel())[0,1]\n",
    "eur_cor = np.corrcoef(np.nan_to_num(np.corrcoef(eur_sub).ravel()), eur_ld.ravel())[0,1]\n",
    "ax[0].text(0.1, 0.9, r'Correlation$ = %.2f$' % afr_cor, fontsize=15, transform=ax[0].transAxes)\n",
    "ax[1].text(0.1, 0.9, r'Correlation$ = %.2f$' % eur_cor, fontsize=15, transform=ax[1].transAxes)\n",
    "\n",
    "cax = ax[1].inset_axes([1.06, 0, 0.05, 1])\n",
    "#cax.yaxis.set_label_position(\"right\")\n",
    "cbar = fig.colorbar(im1, cax=cax)\n",
    "cbar.ax.set_ylabel('Normalized log-distance (bp) between loci', fontsize=14, rotation=270, labelpad=15)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('ld_compare.svg', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52db2d43-0a78-4493-ac3e-c46655f60dee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_migrating_tracts(ts, anc):\n",
    "    neanderthal_id = [p.id for p in ts.populations() if p.metadata['name']==anc][0]\n",
    "    migrating_tracts = []\n",
    "    # Get all tracts that migrated into the neanderthal population\n",
    "    for migration in ts.migrations():\n",
    "        if migration.dest == neanderthal_id:\n",
    "            migrating_tracts.append((migration.left, migration.right))\n",
    "\n",
    "    migrating_tracts = np.array(migrating_tracts) \n",
    "    return migrating_tracts[:,1] - migrating_tracts[:,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccc978e5-9eb5-437e-9f0b-88fc7165e4a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "afr_lengths = get_migrating_tracts(mts_common, 'AFR')\n",
    "eur_lengths = get_migrating_tracts(mts_common, 'EUR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5247c135-727c-4fa7-9cc0-248b62e04d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 2, figsize=(8,4), sharex=True, sharey=True)\n",
    "\n",
    "pt_plot = np.random.binomial(n=1,p=1e-1,size=n_loci*n_loci).astype(bool)\n",
    "\n",
    "ax[0].scatter(np.log10(site_distance.ravel()[pt_plot]/seq_length), np.abs(afr_ld.ravel()[pt_plot]), s=5, label='Local ancestry-adjusted LD correlation', color='red', rasterized=True)\n",
    "ax[0].hist(np.log10((afr_tracts.right-afr_tracts.left) / seq_length), bins=41, density=True, edgecolor='black',histtype=u'step', linestyle='--', linewidth=1.5, label='Local ancestry length distribution')\n",
    "\n",
    "ax[1].scatter(np.log10(site_distance.ravel()[pt_plot]/seq_length), np.abs(eur_ld.ravel()[pt_plot]), s=5, color='red', rasterized=True)\n",
    "ax[1].hist(np.log10((eur_tracts.right-eur_tracts.left) / seq_length), bins=41, density=True, edgecolor='black',histtype=u'step', linestyle='--', linewidth=1.5)\n",
    "\n",
    "fig.text(0.5, -0.02, 'Normalized log-distance (bp) between loci', fontsize=15, ha='center', va='center')\n",
    "\n",
    "ax[0].set_title('African', fontsize=15)\n",
    "ax[1].set_title('European', fontsize=15)\n",
    "\n",
    "\n",
    "handles, labels = ax[0].get_legend_handles_labels()\n",
    "fig.legend(handles, labels, frameon=False, fontsize=15, loc='lower center', bbox_to_anchor=(0.5, 0.97))\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('ld_vs_lanc.svg', bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
